src/streaming/models/CmcdModel.js

```diff
@@ -124,13 +128,29 @@ function CmcdModel() {
         _isStartup = {};
         _initialMediaRequestsDone = {};
         _lastMediaTypeRequest = undefined;
+        _playbackStartedTime = undefined;
+        _msdSent = false;
         _updateStreamProcessors();
     }
 
     function _onPeriodSwitchComplete() {
         _updateStreamProcessors();
     }
 
+    function _onPlaybackStarted() {
+        if (!_playbackStartedTime) {
+            _playbackStartedTime = Date.now();
+        }
+    }
+
+    function _onPlaybackPlaying() {
+        if (!_playbackStartedTime || internalData.msd) {
```

Comment:
I think this is not the right way to measure the media start delay as it does not include MPD parsing time and buffering time. I have reached out to the CMCD experts in the DASH-IF group with the following questions:

> Quick question for the CMCD experts here regarding „Media Start Delay (msd)“. it is defined as in CMCD v2 as:
> 
> „Measures the initial delay in wall-clock time from when a player is instructed to play media for a given session to when any media begins playback“
> 
> •⁠  ⁠I assume this is the time between an application passing the MPD to the player until the player transitions into playing state?
> •⁠  ⁠⁠Or is it the time between a player having enough data in the buffer until the time the player transitions into playing state?
> •⁠  ⁠⁠How is this value derived if a player is instructed to buffer content but not start the playback until a user manually clicks on play? In this case is it the time between passing the MPD and the player having filled the target buffer and being able to start playing once the play button is clicked?
> 
> We have a pending PR by Qualabs that measures the time between the player/app calling the play() method on the video element and the video element dispatching „playing“ state. However that does not include the MPD parsing time and the buffering time. Play() is only called by dash.js when the buffer reaches the initial target value.

Reply:
> •⁠ ⁠I assume this is the time between an application passing the MPD to the player until the player transitions into playing state?

No

> •⁠ ⁠⁠Or is it the time between a player having enough data in the buffer until the time the player transitions into playing state?

Not that either.

We want MSD to measure the user-perceived starting delay. So that should be from when the user expects to see playback, until they see it. For a player without preloading, this will correspond to the time delta between the manifest request and when media is rendering for the user. However for a player with preloading (the example you give), it should measure the time the user manually clicks play to when they see (or hear) the media. For a preloaded player, MSD will will therefore report a value close to zero. That is correct, as it is measuring what the user actually experienced, which was almost instant startup. 

MSD can be more strictly defined, as the spec says, as .. the time difference between the "starting" and "playing" states. A player which is has preloaded but not rendered content is in a "preloading" state. When the user initiates playback, it transitions to "starting" and when media renders (even if the target bufffer has not yet been reached" it transitions to the "playing" state. I tried to represent this in the diagram below:

![Screenshot 2024-11-04 at 14 32 44](https://github.com/user-attachments/assets/7abd4e66-891b-4430-be67-4c3ea60acecf)

Reply:
Thanks @wilaw . @JoaquinBCh Please adjust the implementation to account for what Will described:

* In case `autoplay` is set to false, the player will wait for the playback to be manually triggered by the user. In this case, your current implementation is correct and `msd` corresponds to the delta of `PLAYBACK_PLAYING` and `PLAYBACK_STARTED`
* In case `autoplay` is set to true dash.js will automatically call the `video.play()` function once `settings.streaming.buffer.initialBufferLevel` is reached. However, in this case, we need to include the time it takes to fetch the manifest and reach the buffer target. We can use an event such as `MANIFEST_LOADING_STARTED` and then listen to `PLAYBACK_PLAYING` again. `autoplay` is defined in `StreamController` but you should be able to access it via `cmcdModel.playbackController.getStreamController().getAutoPlay()`. Then you can set the anchor timestamp accordingly.